### ISSUE TITLE
FS: CompleteMultipart: Incorrect behaviour when malformed ETag is supplied.

### ISSUE THREAD

**Issue by hackintoshrao** [minio/minio#1833]
The following unit test reproduces the issue. Check the uncommented case number 13. XL is passing the test with the correct behavior, FS is failing (gist.github.com is down, posting the validation test here). 

```
func TestObjectCompleteMultipartUpload(t *testing.T) {
    ExecObjectLayerTest(t, testObjectCompleteMultipartUpload)
}

// Tests validate CompleteMultipart functionality.
func testObjectCompleteMultipartUpload(obj ObjectLayer, instanceType string, t *testing.T) {

    bucketNames := []string{"minio-bucket", "minio-2-bucket"}
    objectNames := []string{"minio-object-1.txt"}
    uploadIDs := []string{}

    // bucketnames[0].
    // objectNames[0].
    // uploadIds [0].
    // Create bucket before intiating NewMultipartUpload.
    err := obj.MakeBucket(bucketNames[0])
    if err != nil {
        // Failed to create newbucket, abort.
        t.Fatalf("%s : %s", instanceType, err.Error())
    }
    // Initiate Multipart Upload on the above created bucket.
    uploadID, err := obj.NewMultipartUpload(bucketNames[0], objectNames[0], nil)
    if err != nil {
        // Failed to create NewMultipartUpload, abort.
        t.Fatalf("%s : %s", instanceType, err.Error())
    }

    uploadIDs = append(uploadIDs, uploadID)

    // Create multipart parts.
    // Need parts to be uploaded before CompleteMultipart can be called and tested.
    createPartCases := []struct {
        bucketName      string
        objName         string
        uploadID        string
        PartID          int
        inputReaderData string
        inputMd5        string
        intputDataSize  int64
        expectedMd5     string
    }{
        // Case 1-4.
        // Creating sequence of parts for same uploadID.
        {bucketNames[0], objectNames[0], uploadIDs[0], 1, "abcd", "e2fc714c4727ee9395f324cd2e7f331f", int64(len("abcd")), "e2fc714c4727ee9395f324cd2e7f331f"},
        {bucketNames[0], objectNames[0], uploadIDs[0], 2, "efgh", "1f7690ebdd9b4caf8fab49ca1757bf27", int64(len("efgh")), "1f7690ebdd9b4caf8fab49ca1757bf27"},
        {bucketNames[0], objectNames[0], uploadIDs[0], 3, "ijkl", "09a0877d04abf8759f99adec02baf579", int64(len("abcd")), "09a0877d04abf8759f99adec02baf579"},
        {bucketNames[0], objectNames[0], uploadIDs[0], 4, "mnop", "e132e96a5ddad6da8b07bba6f6131fef", int64(len("abcd")), "e132e96a5ddad6da8b07bba6f6131fef"},
    }
    // Iterating over creatPartCases to generate multipart chunks.
    for _, testCase := range createPartCases {
        _, err := obj.PutObjectPart(testCase.bucketName, testCase.objName, testCase.uploadID, testCase.PartID, testCase.intputDataSize,
            bytes.NewBufferString(testCase.inputReaderData), testCase.inputMd5)
        if err != nil {
            t.Fatalf("%s : %s", instanceType, err.Error())
        }
    }

    testCases := []struct {
        bucket   string
        object   string
        uploadID string
        parts    []completePart
        // Expected output of Complete Multipart Upload.
        expectedS3MD5 string
        expectedErr   error
        // Flag indicating whether the test is expected to pass or not.
        shouldPass bool
    }{
        // Test cases with invalid bucket names (Test number 1-4).
        // {".test", "", "", []completePart{}, "", BucketNameInvalid{Bucket: ".test"}, false},
        // {"Test", "", "", []completePart{}, "", BucketNameInvalid{Bucket: "Test"}, false},
        // {"---", "", "", []completePart{}, "", BucketNameInvalid{Bucket: "---"}, false},
        // {"ad", "", "", []completePart{}, "", BucketNameInvalid{Bucket: "ad"}, false},
        // // Test cases for listing uploadID with single part.
        // // Valid bucket names, but they donot exist (Test number 5-7).
        // {"volatile-bucket-1", "", "", []completePart{}, "", BucketNotFound{Bucket: "volatile-bucket-1"}, false},
        // {"volatile-bucket-2", "", "", []completePart{}, "", BucketNotFound{Bucket: "volatile-bucket-2"}, false},
        // {"volatile-bucket-3", "", "", []completePart{}, "", BucketNotFound{Bucket: "volatile-bucket-3"}, false},
        // // Test case for Asserting for invalid objectName (Test number 8).
        // {bucketNames[0], "", "", []completePart{}, "", ObjectNameInvalid{Bucket: bucketNames[0]}, false},
        // // Asserting for Invalid UploadID (Test number 9).
        // {bucketNames[0], objectNames[0], "abc", []completePart{}, "", InvalidUploadID{UploadID: "abc"}, false},
        // // Test case with invalid Part Etag (Test number 10-11).
        // {bucketNames[0], objectNames[0], uploadIDs[0], []completePart{{ETag: "abc"}}, "", fmt.Errorf("encoding/hex: odd length hex string"), false},
        // {bucketNames[0], objectNames[0], uploadIDs[0], []completePart{{ETag: "abcz"}}, "", fmt.Errorf("encoding/hex: invalid byte: U+007A 'z'"), false},
        // // Part number 0 doesn't exist, expecting InvalidPart errori (Test number 12).
        // {bucketNames[0], objectNames[0], uploadIDs[0], []completePart{{ETag: "abcd", PartNumber: 0}}, "", InvalidPart{}, false},
        // Upload and PartNumber exists, But a deliberate Etag mismatch is introduced (Test number 13).
        {bucketNames[0], objectNames[0], uploadIDs[0], []completePart{{ETag: "abcd", PartNumber: 1}}, "", BadDigest{}, false},
        // Test case for uploadID with multiple parts (Test number 14).
        // {bucketNames[0], objectNames[0], uploadIDs[0], 0, 10, partInfos[0], nil, true},
        // Test case with maxParts set to less than number of parts (Test number 15).
        // {bucketNames[0], objectNames[0], uploadIDs[0], 0, 3, partInfos[1], nil, true},
        // Test case with partNumberMarker set (Test number 16).
        // {bucketNames[0], objectNames[0], uploadIDs[0], 3, 2, partInfos[2], nil, true},
    }

    for i, testCase := range testCases {
        actualResult, actualErr := obj.CompleteMultipartUpload(testCase.bucket, testCase.object, testCase.uploadID, testCase.parts)
        if actualErr != nil && testCase.shouldPass {
            t.Errorf("Test %d: %s: Expected to pass, but failed with: <ERROR> %s", i+1, instanceType, actualErr.Error())
        }
        if actualErr == nil && !testCase.shouldPass {
            t.Errorf("Test %d: %s: Expected to fail with <ERROR> \"%s\", but passed instead", i+1, instanceType, testCase.expectedErr.Error())
        }
        // Failed as expected, but does it fail for the expected reason.
        if actualErr != nil && !testCase.shouldPass {
            if !strings.Contains(actualErr.Error(), testCase.expectedErr.Error()) {
                t.Errorf("Test %d: %s: Expected to fail with error \"%s\", but instead failed with error \"%s\"", i+1, instanceType, testCase.expectedErr.Error(), actualErr.Error())
            }
        }
        // Passes as expected, but asserting the results.
        if actualErr == nil && testCase.shouldPass {

            // Asserting s3MD%.
            if actualResult != testCase.expectedS3MD5 {
                t.Errorf("Test %d: %s: Expected the result to be \"%v\", but found it to \"%v\"", i+1, instanceType, testCase.expectedS3MD5, actualResult)
            }
        }
    }
}

```

**Comment by krishnasrinivas** [comment]
looking at this.

**Comment by krishnasrinivas** [comment]
fixed by 67bba270a056ad6659c04585e283c07c17ac9a57
